python/dolfinx/fem/nonlinearproblem.py

```diff
@@ -0,0 +1,125 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2021 Jørgen S. Dokken
+#
+# This file is part of DOLFINX (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+
+from typing import List
+import ufl
+from petsc4py import PETSc
+from dolfinx import fem, cpp
+import mpi4py
+
+
+class NonlinearProblem:
+    """Nonlinear problem class for solving the non-linear problem
+    F(u, v) = 0 for all v in V
+
+    """
+
+    def __init__(self, F: ufl.form.Form, u: fem.Function, bcs: List[fem.DirichletBC] = [], J: ufl.form.Form = None,
+                 form_compiler_parameters={}, jit_parameters={}):
+        """Initialize class that sets up structures for solving the non-linear problem using Newton's method,
+        dF/du(u) du = -F(u)
+
+        Parameters
+        ----------
+        F
+            The PDE residual F(u, v)
+        u
+            The unknown
+        bcs
+            List of Dirichlet boundary conditions
+        J
+            UFL representation of the Jacobian (Optional)
+        form_compiler_parameters
+            Parameters used in FFCX compilation of this form. Run `ffcx --help` at
+            the commandline to see all available options. Takes priority over all
+            other parameter values, except for `scalar_type` which is determined by
+            DOLFINX.
+        jit_parameters
+            Parameters used in CFFI JIT compilation of C code generated by FFCX.
+            See `python/dolfinx/jit.py` for all available parameters.
+            Takes priority over all other parameter values.
+
+        .. code-block:: python
+            problem = LinearProblem(F, u, [bc0, bc1])
+        """
+        self.L = fem.form.Form(F, form_compiler_parameters=form_compiler_parameters, jit_parameters=jit_parameters)
```

Comment:
Can you make `self.L, self.a` just read only private properties? The same as in `LinearProblem`.

Reply:
@jorgensd  I meant to make it exactly as in `LinearProblem`, so add property getters for private variables. Private viariables `_foo` should be accessed only via getters. It is more clean and makes it clear what is private/public API.